### PR TITLE
[ADD][9.0] stock_picking_package_info

### PR DESCRIPTION
**Pull request by lasley**
This module displays in the picking_ids, all packages appearing in operations and
the package provides the following information:
- Width
- Height
- Length
- Total estimated weight
- Total estimated net weight
- Permitted volume
- Total volume charge
- Empty package weight
- Real weight

Logic was provided by @odoomrp, just upgraded to v9

Few notes, making this a WIP if someone will please tag
- Save for later method was removed & logic for `create_all_move_packages` was moved to the `process` method of the wizard. From that I can tell, this allows us to not need to save due to more streamlined v9 wizard, but I could be wrong and would love a comment from @pedrobaeza  on that 😉 
- Tests are failing, and more need to be added

### CODE REVIEW

**Comment by lasley** [comment]
@pedrobaeza - Can you assist with the missing CLA? I don't want to bother fixing up all the tests and stuff until I know we're good in that respect 😉

**Comment by pedrobaeza** [comment]
It's covered by Avanzosc's CLA. cc @anajuaristi

**Comment by lasley** [comment]
Hooray thanks @pedrobaeza ! I'll swoop these tests up in the next few days then

File: .travis.yml
Comment (lasley) [inline .travis.yml:44]: Just to bring attention to this - we will remove before merge, it's just so we can identify the coverage gaps while working. A rebase onto current 9 contained a not fun amount of conflict work, so this seemed like a good interim solution.

**Comment by lasley** [comment]
@t3ddftw did you have to make any changes to this while functionally testing during implementation in laslabs/odoo-connector-easypost#2?

**Comment by tedsalmon** [comment]
Hey @lasley,

Yes, these are the changes I made:
- Add name_get to `product.packaging.template` to return dimensions in name
- Fix spelling of 'excluding' in help of `stock.quant` -> `weight_net`
- Remove requirement of `product_pack_tmpl_id` in `stock.quant.package`
- Add Packaging Template to product_packaging_view
- Update invisible attribute in stock_picking_view.xml to use `package_ids` instead of outdated `package` attribute

**Comment by lasley** [comment]
Excellent, please add relevant logic into this PR so we can review 😄

**Comment by tedsalmon** [comment]
@lasley,

The changes have already been dropped in and squashed :)

File: .travis.yml
Comment (lasley) [reply .travis.yml:44]: Rebase onto current 9.0 so this commit is removed, PR was merged that included it

File: stock_picking_package_info/__openerp__.py
Comment (lasley) [inline stock_picking_package_info/__openerp__.py]: Change this to Tecnativa - correct me if I'm wrong @pedrobaeza

File: stock_picking_package_info/models/product_packaging.py
Comment (lasley) [inline stock_picking_package_info/models/product_packaging.py]: This string is implicit, remove

File: stock_picking_package_info/models/product_packaging.py
Comment (lasley) [inline stock_picking_package_info/models/product_packaging.py]: s/layer/Layer

File: stock_picking_package_info/models/product_packaging_template.py
Comment (lasley) [inline stock_picking_package_info/models/product_packaging_template.py]: select is deprecated, `index=True`

File: stock_picking_package_info/models/stock_picking.py
Comment (lasley) [inline stock_picking_package_info/models/stock_picking.py]: This also depends on `pack_operation_ids.package_id`

File: stock_picking_package_info/models/stock_picking.py
Comment (lasley) [inline stock_picking_package_info/models/stock_picking.py]: This also depends on `pack_operation_ids.package_id`, `pack_operation_ids.quant_ids`

File: stock_picking_package_info/models/stock_picking.py
Comment (lasley) [inline stock_picking_package_info/models/stock_picking.py]: This attr is established to exist in the if statement. It would be cleaner and more efficient to use dict lookup here instead of .get

File: stock_picking_package_info/models/stock_picking.py
Comment (lasley) [inline stock_picking_package_info/models/stock_picking.py]: This attr is established to exist in the if statement. It would be cleaner and more efficient to use dict lookup here instead of .get

File: stock_picking_package_info/models/stock_picking.py
Comment (lasley) [inline stock_picking_package_info/models/stock_picking.py]: Dict assignment would be more efficient here

File: stock_picking_package_info/models/stock_picking_package_weight_lot.py
Comment (lasley) [inline stock_picking_package_info/models/stock_picking_package_weight_lot.py]: Remove blank line

File: stock_picking_package_info/models/stock_quant_package.py
Comment (lasley) [inline stock_picking_package_info/models/stock_quant_package.py]: Bit heavy on the logging, let's combine these statements

File: stock_picking_package_info/reports/label_creator_palet_report.xml
Comment (lasley) [inline stock_picking_package_info/reports/label_creator_palet_report.xml]: Use `<odoo>` tag & remove data

File: stock_picking_package_info/views/product_packaging_template_view.xml
Comment (lasley) [inline stock_picking_package_info/views/product_packaging_template_view.xml]: Use new header & remove perpetual copyright. Also, did we create this file? Shouldn't there be another entity copyright here?

File: stock_picking_package_info/views/product_packaging_template_view.xml
Comment (lasley) [inline stock_picking_package_info/views/product_packaging_template_view.xml]: Remove data tag

File: stock_picking_package_info/views/product_packaging_view.xml
Comment (lasley) [inline stock_picking_package_info/views/product_packaging_view.xml]: Use new header & remove perpetual copyright. Also, did we create this file? Shouldn't there be another entity copyright here?

File: stock_picking_package_info/views/product_packaging_view.xml
Comment (lasley) [inline stock_picking_package_info/views/product_packaging_view.xml]: Use odoo tag, remove data

File: stock_picking_package_info/views/product_template_view.xml
Comment (lasley) [inline stock_picking_package_info/views/product_template_view.xml]: Use odoo tag

File: stock_picking_package_info/views/product_template_view.xml
Comment (lasley) [inline stock_picking_package_info/views/product_template_view.xml]: Use new header & remove perpetual copyright. Also, did we create this file? Shouldn't there be another entity copyright here?

File: stock_picking_package_info/views/stock_picking_view.xml
Comment (lasley) [inline stock_picking_package_info/views/stock_picking_view.xml]: Use new header & remove perpetual copyright. Also, did we create this file? Shouldn't there be another entity copyright here?

File: stock_picking_package_info/views/stock_picking_view.xml
Comment (lasley) [inline stock_picking_package_info/views/stock_picking_view.xml]: odoo tag, remove data

File: stock_picking_package_info/views/stock_quant_package_view.xml
Comment (lasley) [inline stock_picking_package_info/views/stock_quant_package_view.xml]: Use new header & remove perpetual copyright. Also, did we create this file? Shouldn't there be another entity copyright here?

File: stock_picking_package_info/views/stock_quant_package_view.xml
Comment (lasley) [inline stock_picking_package_info/views/stock_quant_package_view.xml]: odoo tag, remove data

**Review by lasley — COMMENTED** [review]
Thanks @t3ddftw & @bmw95 - Comments inline

File: stock_picking_package_info/models/stock_quant_package.py
Comment (woodbrettm) [reply stock_picking_package_info/models/stock_quant_package.py]: I think those may have been added for testing purposes. Can probably just be removed.

File: stock_picking_package_info/views/product_packaging_template_view.xml
Comment (tedsalmon) [reply stock_picking_package_info/views/product_packaging_template_view.xml]: For the record, these files did not have a copyright on them previously. [See this link.](https://github.com/odoomrp/odoomrp-wip/blob/8.0/stock_picking_package_info/views/stock_quant_package_view.xml)

File: .travis.yml
Comment (tedsalmon) [reply .travis.yml:44]: This is actually caught up with 9.0 already.

**Comment by tedsalmon** [comment]
Hey @lasley,

All the requested changes have been made and squashed.

Thanks! 😄

**Comment by lasley** [comment]
@t3ddftw - Please don't squash your changes after reviews. I have to review all of this code over again now (35 files), instead of just the diff.

File: .travis.yml
Comment (lasley) [reply .travis.yml:44]: You didn't rebase onto 9.0 like I asked. Please do. This is in the diff & it should not be

**Comment by lasley** [comment]
@t3ddftw - please look into Runbot warnings:

```
2016-10-19 15:05:39,859 3050 WARNING 3270174-247-066d5a-all openerp.models.schema: Table `product_packaging`: column `qty` has changed type (DB=float8, def=integer), data moved to column `qty_moved0`
2016-10-19 15:05:40,011 3050 INFO 3270174-247-066d5a-all openerp.sql_db: Programming error: column "name" named in key does not exist
, in query ALTER TABLE "product_packaging_template" ADD CONSTRAINT "product_packaging_template_name_uniq" UNIQUE(name)
2016-10-19 15:05:40,012 3050 WARNING 3270174-247-066d5a-all openerp.models.schema: Table 'product_packaging_template': unable to add 'UNIQUE(name)' constraint !
 If you want to have it, you should update the records and execute manually:
ALTER TABLE "product_packaging_template" ADD CONSTRAINT "product_packaging_template_name_uniq" UNIQUE(name)
```

**Comment by tedsalmon** [comment]
Runbot warnings have been fixed. I have also updated the tests for stock_picking_package_preparation to be compatible with this module.

Thanks!

**Comment by lasley** [comment]
Runbot 🔴 

```
 _auto_init Table `product_packaging`: column `qty` has changed type (DB=float8, def=integer), data moved to column `qty_moved0`
```

**Comment by tedsalmon** [comment]
@lasley,

Fixed - my apologies, I thought I had knocked that one out. All warnings gone and all checks are green.

Thanks!

**Review by lasley — COMMENTED** [review]
Thanks @t3ddftw 👍

File: stock_picking_package_info/__openerp__.py
Comment (pedrobaeza) [inline stock_picking_package_info/__openerp__.py]: Please put only `Tecnativa`

File: stock_picking_package_info/models/product_packaging_template.py
Comment (pedrobaeza) [inline stock_picking_package_info/models/product_packaging_template.py]: Don't use a Python reserved word. Please use something like `tmpl_type`

File: stock_picking_package_info/models/product_product.py
Comment (pedrobaeza) [inline stock_picking_package_info/models/product_product.py]: We should have this weight at product variant level, not at template level, but when setting at template level, it should be propagated to the variants.

**Review by pedrobaeza — CHANGES_REQUESTED** [review]
As the module also provides information for quants and product, it should be renamed to `stock_package_info`

File: stock_picking_package_info/models/product_packaging_template.py
Comment (lasley) [reply stock_picking_package_info/models/product_packaging_template.py]: IMO abbreviations should be avoided too.

FYI type is not a reserved word, it's a builtin function - https://docs.python.org/3/reference/lexical_analysis.html#keyword

Comment (pedrobaeza) [reply stock_picking_package_info/models/product_packaging_template.py]: `tmpl` is a common abbreviation you find across Odoo, but OK then to put `template_type`. The word is not a keyword, but a function name that is in the base: http://stackoverflow.com/questions/10568087/is-it-safe-to-use-the-python-word-type-in-my-code

Comment (lasley) [reply stock_picking_package_info/models/product_packaging_template.py]: Builtin methods are technically a part of the `__builtins__` namespace though, so aren't reserved in all namespaces like keywords.

That said, I'm good with the more descriptive term. I think `package_type` might fit better in the context though, but we can leave that to @t3ddftw to figure out 😛

File: stock_picking_package_info/__openerp__.py
Comment (anajuaristi) [reply stock_picking_package_info/__openerp__.py]: @pedrobaeza
Why only tecnativa if module was developed by odoomrp team and avanzosc.
Maybe you mean to erase your own name from authors since now your company is Tecnativa. Is this? 
Thank you for clarifying

Comment (pedrobaeza) [reply stock_picking_package_info/__openerp__.py]: Of course, Ana. I'm talking about replacing `Tecnativa - Pedro M. Baeza` by only `Tecnativa`. No need to be picky about this.

Comment (anajuaristi) [reply stock_picking_package_info/__openerp__.py]: As I thougth on the first term.
It was just to clarify because your request was a bit confusing.
Thank you!!

File: stock_package_info/models/product_product.py
Comment (tedsalmon) [inline stock_package_info/models/product_product.py]: I'm open to suggestions here as this seems a bit hacky to me. This is my fix for the infinite recursion caused by the fact that the given product is itself a variant of its variants.

Comment (woodbrettm) [reply stock_package_info/models/product_product.py]: I came across some code implemented for a similar issue when migrating [medical_pathology](https://github.com/OCA/vertical-medical/blob/9.0/medical_pathology/models/medical_pathology_category.py). Not sure if that'll help though in this case.

File: stock_package_info/models/product_product.py
Comment (lasley) [inline stock_package_info/models/product_product.py]:
```
for record in self:
    variants = record.mapped('product_variant_ids').filtered(
        lambda r: r.id != record.id
    )
    variants.write({...
```

Will solve your infinite recursion

Comment (pedrobaeza) [reply stock_package_info/models/product_product.py]: You can omit `id` part:

```
    variants = record.mapped('product_variant_ids').filtered(
        lambda r: r != record
    )
```

**Review by lasley — COMMENTED** [review]
Thanks for the updates @t3ddftw 👍

File: stock_package_info/models/product_packaging_template.py
Comment (lasley) [inline stock_package_info/models/product_packaging_template.py:31]: Turns out this col name can cause issues. See https://github.com/odoo/odoo/issues/13780#issuecomment-260872554

Comment (pedrobaeza) [reply stock_package_info/models/product_packaging_template.py:31]: No, that's for JS, not Python.

Comment (lasley) [reply stock_package_info/models/product_packaging_template.py:31]: Oh got it, thanks @pedrobaeza - I was thinking that the field names would transfer over to JS at some point in a widget or something. Sweet, this was kind of worrisome.

Comment (pedrobaeza) [reply stock_package_info/models/product_packaging_template.py:31]: Oh, wait, my fault. That's exactly the problem... so you have to change it indeed.

Comment (lasley) [reply stock_package_info/models/product_packaging_template.py:31]: Although I'm wondering if we should bite the bullet in this particular instance. Renaming this will create a conflict with the `product.packaging` model that it is templating- https://github.com/odoo/odoo/blob/fc2e80cb4bcc450762c7ac5cb82a3e2d88062b38/addons/delivery/models/product_packaging.py#L12

Comment (lasley) [reply stock_package_info/models/product_packaging_template.py:31]: Although we should change these all to integer fields in order to match the types in the model. Carriers don't accept fractions of measurements on the dimensions anyways.

Comment (pedrobaeza) [reply stock_package_info/models/product_packaging_template.py:31]: This happens on one2many lists that contains a column with this name. Maybe we can avoid the problem if this is not the case.

Comment (lasley) [reply stock_package_info/models/product_packaging_template.py:31]: Sweet yeah I think we'll be safe here then as I don't see us presenting this data in that format.

Comment (tedsalmon) [reply stock_package_info/models/product_packaging_template.py:31]: Hey all,

In https://github.com/OCA/stock-logistics-workflow/pull/247/commits/8151c7b882fe48911e6995e993e383baf7d4ab29 I updated all the measurement field names by prepending `pack_` to them as I thought only updating the `length`  field name would create a weird incongruity in naming. Does anyone oppose this change?

Thanks!

**Comment by lasley** [comment]
@t3ddftw - Are we good for a final review on this?

File: stock_package_info/__init__.py
Comment (lmignon) [inline stock_package_info/__init__.py:8]: @lasley You comment the import  but the wizard still exist into the code.

**Review by lmignon — CHANGES_REQUESTED** [review]

**Comment by lasley** [comment]
So this one doesn't really work that well, and will require a lot of work IMO. We're abandoning for a more simple approach that fits our needs of being able to dispatch things with a package.

The main reason for my thoughts here is that too many concepts are being blended into `product.packaging` and `stock.quant.packaging`. It complicates everything, and really doesn't play nicely with pre-existing databases - as indicated by the test failures here about the UOMs.

We will be losing a lot of the volume calculation and stuff here, but that's out of scope for my needs - sorry. Our new solution is in https://github.com/OCA/delivery-carrier/pull/141

**Pull request closed without merging**
